### PR TITLE
Fix s_copy_ and indexing types

### LINKED ISSUE #592: type promotion
```
import torch
import torch_xla
import torch_xla_py.utils as xu
import torch_xla_py.xla_model as xm
import pdb
import torch.nn.functional as F
xla_device = xm.xla_device()
# xla_device = 'cpu'

pdb.set_trace()

a = torch.rand(10).to(xla_device).to(dtype=torch.uint8)
b = torch.tensor([0, 1, 2, 3]).to(xla_device)
print(a.dtype)
print(a[b].dtype) # shows float32, should be uint8

### PR DESCRIPTION
**Pull request by asuhan**
We need to cast to the destination type on the same device path.
Fixes #592.

### CODE REVIEW

File: test/cpp/cpp_test_util.cpp [hunk 1]
```diff
@@ -25,11 +46,13 @@ at::Tensor ToCpuTensor(const at::Tensor& t) {
   return xtensor ? ToTensor(*xtensor) : tensor;
 }
 
-bool EqualValues(at::Tensor tensor1, at::Tensor tensor2) {
-  if (tensor1.sizes() != tensor2.sizes()) {
-    std::cerr << "Different sizes:\n"
-              << tensor1.sizes() << "\n-vs-\n"
-              << tensor2.sizes() << "\n";
+bool EqualValues(at::Tensor tensor1, at::Tensor tensor2,
```
Comment (dlibenzi) [inline test/cpp/cpp_test_util.cpp]: Do not overload this one. Add a new one with proper name, that also does the element type check.

File: test/cpp/cpp_test_util.cpp [hunk 2]
```diff
@@ -66,6 +90,16 @@ bool CloseValues(at::Tensor tensor1, at::Tensor tensor2, double rtol,
   return equal;
 }
 
+at::Tensor MakeRandomTensor(at::IntArrayRef size, at::ScalarType scalar_type) {
```
Comment (dlibenzi) [inline test/cpp/cpp_test_util.cpp]: Why?
We had one generator and we removed it since it was obfuscating the code.
Leave at::rand() alone, and if we really need and int-rand, add an helper. I don't think we need that.

File: torch_xla/csrc/tensor_methods.cpp [hunk 3]
```diff
@@ -1600,11 +1607,14 @@ XLATensor XLATensor::rsub(const XLATensor& input, at::Scalar other,
 
 void XLATensor::s_copy_(XLATensor& input, XLATensor& src) {
   if (input.GetDevice() == src.GetDevice()) {
-    input.SetIrValue(src.GetIrValue());
+    input.SetIrValue(
```
Comment (dlibenzi) [inline torch_xla/csrc/tensor_methods.cpp]: I would avoid adding a cast IR node if it is not needed.
So I would check for src dtype and our dtype, and avoid that.
We are also missing the other path in aten_xla_type.cpp line 2347. We need to possibly convert the tensor (only if dtype differs) before a SetTensor().

Comment (asuhan) [reply torch_xla/csrc/tensor_methods.cpp]: Right, haven't noticed the other path in `AtenXlaType`.

Regarding the redundant cast, I'm pretty sure we optimize away no-op casts in the XLA back-ends. Doing manual optimizations kinda defeats the purpose of a multi-tier compiler.

File: test/cpp/cpp_test_util.cpp (hunk 2 above)
Comment (asuhan) [reply test/cpp/cpp_test_util.cpp]: I had to do this because we have different random factories depending on the `scalar_type`. The tests iterate through all `scalar_type`s, so a factory with an unique signature was needed. What is your suggestion?

File: test/cpp/cpp_test_util.cpp (hunk 1 above)
Comment (asuhan) [reply test/cpp/cpp_test_util.cpp]: Sounds good, the code duplication would be minor.

File: test/cpp/cpp_test_util.cpp (hunk 2 above)
Comment (dlibenzi) [reply test/cpp/cpp_test_util.cpp]: I would just have us call at::rand() and at::randint() directly.
In the places where scalar type iteration is needed, and only there, we may have a wrapper if we really need to explore values up to max().

File: torch_xla/csrc/tensor_methods.cpp (hunk 3 above)
Comment (dlibenzi) [reply torch_xla/csrc/tensor_methods.cpp]: I know we can. XLA will too. But we better not have graphs polluted with noops, as they are already big enough to inspect for debug.

File: test/cpp/cpp_test_util.cpp (hunk 1 above)
Comment (asuhan) [reply test/cpp/cpp_test_util.cpp]: Done it the other way around since we almost always want to check the element type, only cast operations are different.

File: test/cpp/cpp_test_util.cpp (hunk 2 above)
Comment (asuhan) [reply test/cpp/cpp_test_util.cpp]: Done.

File: torch_xla/csrc/tensor_methods.cpp (hunk 3 above)
Comment (asuhan) [reply torch_xla/csrc/tensor_methods.cpp]: Done.

**Comment by asuhan** [comment]
Failure unrelated and fixed by https://github.com/pytorch/xla/pull/594.

File: torch_xla/csrc/torch_util.h [hunk 4]
```diff
@@ -23,6 +25,13 @@ static inline at::Tensor CopyTensor(const at::Tensor& ref) {
   return ref.to(ref.options(), /*non_blocking=*/false, /*copy=*/true);
 }
 
+// Same as above, with an additional cast.
+static inline at::Tensor CopyAndCastTensor(const at::Tensor& ref,
```
Comment (dlibenzi) [inline torch_xla/csrc/torch_util.h]: I'd just call this CopyTensor() as well.

File: torch_xla/csrc/tensor_methods.cpp [hunk 5]
```diff
@@ -1094,8 +1098,9 @@ XLATensor XLATensor::index_put(
     tensorflow::gtl::ArraySlice<const XLATensor> indices,
     const XLATensor& values, bool accumulate,
     tensorflow::gtl::ArraySlice<const xla::int64> result_permutation) {
-  return input.CreateFrom(IndexPutByTensors(input, indices, values, accumulate,
-                                            result_permutation));
+  return input.CreateFrom(
```
Comment (dlibenzi) [inline torch_xla/csrc/tensor_methods.cpp]: This should be the same as well.
```
x.CreateFrom(y);
```
Expands to:
```
x.CreateFrom(y, x.GetDevice(), x.dtype());
```

Same thing in the other places.

File: torch_xla/csrc/tensor_methods.cpp [hunk 6]
```diff
@@ -1109,10 +1114,12 @@ void XLATensor::index_put_(
 
 XLATensor XLATensor::index_select(const XLATensor& input, xla::int64 dim,
                                   const XLATensor& index) {
-  return input.CreateFrom(ir::MakeNode<ir::ops::IndexSelect>(
-      input.GetIrValue(),
-      XlaHelpers::GetCanonicalDimensionIndex(dim, input.shape().get().rank()),
-      index.GetIrValue()));
+  return input.CreateFrom(
```
[the remark this answers is not in the document]
Comment (asuhan) [reply torch_xla/csrc/tensor_methods.cpp]: Done.

File: torch_xla/csrc/tensor_methods.cpp [hunk 7]
```diff
@@ -1600,11 +1607,18 @@ XLATensor XLATensor::rsub(const XLATensor& input, at::Scalar other,
...
+    } else {
+      input.SetIrValue(
+          ir::MakeNode<ir::ops::Cast>(src.GetIrValue(), input.dtype()));
+    }
   } else {
     // TODO: This can be optimized via proper XRT/XLA computation.
-    XLATensor new_tensor = XLATensor::Create(src.ToTensor(), input.GetDevice(),
-                                             src.RequiresGrad());
+    XLATensor new_tensor =
+        XLATensor::cast(XLATensor::Create(src.ToTensor(), input.GetDevice(),
```
[the remark this answers is not in the document]
Comment (asuhan) [reply torch_xla/csrc/tensor_methods.cpp]: Done.

File: torch_xla/csrc/torch_util.h (hunk 4 above)
Comment (asuhan) [reply torch_xla/csrc/torch_util.h]: Done.

File: torch_xla/csrc/tensor_methods.cpp (hunk 5 above)
Comment (asuhan) [reply torch_xla/csrc/tensor_methods.cpp]: Done.
